storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java

```diff
@@ -0,0 +1,2091 @@
+                CompressionType.NONE,
+                Compression.NONE,
+                false,
+                RecordBatch.MAGIC_VALUE_V0,
+                TimestampType.CREATE_TIME,
+                5000L,
+                5000L,
+                RecordBatch.NO_PARTITION_LEADER_EPOCH,
+                AppendOrigin.CLIENT,
+                MetadataVersion.latestTesting()
+        ).validateMessagesAndAssignOffsets(
+                PrimitiveRef.ofLong(offset), metricsRecorder, RequestLocal.withThreadConfinedCaching().bufferSupplier()
+        ).validatedRecords, offset);
+    }
+
+
+    @Test
+    public void testOffsetAssignmentAfterDownConversionV2ToV0Compressed() {
+        long offset = 1234567;
+        long now = System.currentTimeMillis();
+        Compression compression = Compression.gzip().build();
+        MemoryRecords records = createRecords(RecordBatch.MAGIC_VALUE_V2, now, compression);
+        checkOffsets(records, 0);
+        checkOffsets(new LogValidator(
+                records,
+                new TopicPartition("topic", 0),
+                time,
+                CompressionType.GZIP,
+                compression,
+                false,
+                RecordBatch.MAGIC_VALUE_V0,
+                TimestampType.CREATE_TIME,
+                5000L,
+                5000L,
+                RecordBatch.NO_PARTITION_LEADER_EPOCH,
+                AppendOrigin.CLIENT,
+                MetadataVersion.latestTesting()
+        ).validateMessagesAndAssignOffsets(
+                PrimitiveRef.ofLong(offset), metricsRecorder, RequestLocal.withThreadConfinedCaching().bufferSupplier()
+        ).validatedRecords, offset);
+    }
+
+    @Test
+    public void testNonIncreasingOffsetRecordBatchHasMetricsLogged() {
+        MemoryRecords records = createNonIncreasingOffsetRecords(
+                RecordBatch.MAGIC_VALUE_V2, RecordBatch.NO_TIMESTAMP, Compression.NONE);
+        records.batches().iterator().next().setLastOffset(2);
+        assertThrows(InvalidRecordException.class, () -> new LogValidator(
+                records,
+                new TopicPartition("topic", 0),
+                time,
+                CompressionType.GZIP,
+                Compression.gzip().build(),
+                false,
+                RecordBatch.MAGIC_VALUE_V0,
+                TimestampType.CREATE_TIME,
+                5000L,
+                5000L,
+                RecordBatch.NO_PARTITION_LEADER_EPOCH,
+                AppendOrigin.CLIENT,
+                MetadataVersion.latestTesting()
+        ).validateMessagesAndAssignOffsets(
+                PrimitiveRef.ofLong(0L), metricsRecorder, RequestLocal.withThreadConfinedCaching().bufferSupplier()
+        ));
+        assertEquals(metricsKeySet.stream().filter(m -> m.getMBeanName().endsWith(BrokerTopicStats.InvalidOffsetOrSequenceRecordsPerSec())).count(), 1);
+        assertTrue(kafka.utils.TestUtils$.MODULE$.meterCount(BrokerTopicStats.InvalidOffsetOrSequenceRecordsPerSec()) > 0);
```

Comment:
Could you please avoid using scala TestUtils?

Reply:
Maybe we can create a impl of `MetricsRecorder` to collect the function call?

Reply:
Done. PTAL.